Enh verbose and no execute options

hardikpnsp opened the pull request:
### Enhancement

- Added option `-v` for verbose where all curl and tar commands that are executed are echoed.
- Added option `-n` for no-execute where nothing is executed but if `-v` is passed it will still echo the commands to be executed. 
- Added new test cases for the above options. 

### Checks

- [x] all tests pass 
- [x] docs updated

Closes issue: (#10)

hardikpnsp committed:
ENH: add option -v (verbose) and -n (no-execute)

hardikpnsp committed:
TST: Test cases for -v and -n options

olav35 commented:
Thanks for the PR.

I think it's best if we put the flags right before or after the subcommands(file/url/shorten) and prior to the file/url.

I will push a commit that updates the manpage to describe the desired usage.

olavfosse committed:
Add n and v flags to SYNOPSIS

olavfosse committed:
Describe options in DESCRIPTION

olav35 commented:
It is probably worth looking into GETOPT(1) for parsing the options.

hardikpnsp committed:
ENH: getopt used for more predictable output.
- Flags are now between command_modifier and target

hardikpnsp committed:
Merge branch 'master' of github.com:hardikpnsp/0x0 into enh-verbose-and-no-execute-options

hardikpnsp commented:
used getopt for getting the options
now it works as:
```bash
0x0 file [OPTIONS] filename  
```
there were conflicts in the pdf file, as it's a binary file I ended up using the one in the current branch.

olavfosse committed:
Add options to usage message

olav35 commented:
> there were conflicts in the pdf file, as it's a binary file I ended up using the one in the current branch.

Yeah, I the way to resolve that is just to resolve doc/0x0.1 and generate a new pdf with bin/doc.sh. That script uses mandoc though which you might be missing. Don't worry about it, I'll handled the merge conflicts :).

olav35 commented:
Why is it that the curl commands have a closing parentheses appended to them?

olavfosse committed:
Clean up files and directories for each test

olavfosse committed:
Unset all local* variables per test

This commit, along with the previous one should ensure that each
test is entirely self sufficient with regards to the local*
variables and files they used decoupling the tests from each other.

* Local in the sense that they are only to be used within the test
  in question

olavfosse committed:
Fix indention

hardikpnsp commented:
> Why is it that the curl commands have a closing parentheses appended to them?

That seems like a typo, let me check.

olav35 commented:
To DRY the code we should use a function like `maybe_run_and_maybe_print`, or something like that, instead of WET.

olavfosse commented on src/0x0:
```diff
@@ -80,17 +83,22 @@ request() {
-		*)
-			echo "$response"
-	esac
+	if $verbose ; then
+		echo "\$ curl -Ss -w 'status_code=%{http_code}' https://0x0.st "$form_string")"
+	fi
+	if ! $no_execute ; then
```
WET

olavfosse commented on src/0x0:
```diff
@@ -104,7 +112,14 @@ dispatch_file() {
 		# stderr is redirected to /dev/null to prevent the following from being displayed
 		# tar: Removing leading / from absolute path names in the archive
-		tar cf - "$file" 2> /dev/null | request file - || exit_with_error 'error: tar archival failed'
+		if $verbose ; then
+			echo "\$ tar cf - "$file" 2> /dev/null"
+		fi
+		if ! $no_execute ; then
```
WET

hardikpnsp committed:
FIX: removed parenthesis (typo) when printing curl

olav35 commented:
When I think about it, using a function like `maybe_run_and_maybe_print` is not as straight forward as I initially thought. This is because the output of the curl command is further processed and separating the command (to be displayed to the user if -v) and the output of the command sounds quite hackish. We can just assign the command to a variable and print/run that instead.

olavfosse committed:
Replace no_execute with execute(remove double negation)

olavfosse committed:
Add test 18

olavfosse committed:
Refactor dispatch

olavfosse committed:
Add test 19

olavfosse committed:
Pass test 19

olavfosse committed:
Remove redundant guard clauses from the dispatch handlers

olavfosse committed:
Refactor to avoid global* execute and verbose variables

* Technically they are still global, but now they are treated as
  locals and passed down as arguments.

olavfosse commented on src/0x0:
```diff
@@ -77,57 +77,73 @@ form_string() {
+		fi
+		if "$execute"; then
+			tar cf - "$file" 2> /dev/null | request file - "$execute" "$verbose" || exit_with_error 'error: tar archival failed'
+		else
+			request file - "$execute" "$verbose"
+		fi
```
?

olavfosse replied:
My bad, we need to run it in case verbose is enabled.

olavfosse replied:
I thought, it was redundant at first. I added a comment that says it's not.

olavfosse committed:
Add a NB comment

olavfosse committed:
Change how the commands are outputted in verbose mode

olavfosse committed:
Remove redundant quoting

olavfosse committed:
Remove pointless /dev/null redirection

olavfosse committed:
Order command modifiers alphabetically

olavfosse committed:
Add -nv example to EXAMPLE

olav35 commented:
Tests are passing on MacOS, OpenBSD and Ubuntu. Merging.
